### PR TITLE
`Programming exercises`: Allow instructors to create assignment repository after due date passed

### PR DESCRIPTION
**Pull request by basak-akan**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).

#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

Resolves [#6430](https://github.com/ls1intum/Artemis/issues/6430)
When trying to create an assignment repository for an instructor after the due date of the programming exercise has passed, an error is thrown. Instructors should be able to create an assignment repository for a programming exercise even after the exercise due date has passed.

### Description

Participation in programming exercises can only be started if the participation feature is enabled and the due date has not yet passed. However, instructors and editors can still trigger create an assignment repository even after the due date. To address this,  it is implemented a verification check in the participation initiation process. This check ensures that the user's role is at least that of an editor. For more detailed information on access rights related to programming exercises, please refer to our [Access Rights](https://docs.artemis.cit.tum.de/admin/accessRights/#programming-exercises).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Editor
- 1 Programming Exercise with due date passed

1. Log in to Artemis
2. Go to Course Management and look for the exercise with the due date in the past as an instructor.
3. Click on 'Edit in editor'
4. Click on 'Create assignment repository'

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ParticipationResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5102-6/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ParticipationResource.html) | 91% | ✅ |

### Screenshots

For Instructors 
image

After the assignment repository is created Delete Assignment Repository button is visible and enabled
image

For Editors 
image

After the assignment repository is created Delete Assignment Repository button is not visible and the feature is not enabled due to access rights. Once participation is triggered, editors do not have permission to delete.
image

### CODE REVIEW

**Review by JohannesStoehr — APPROVED** [review]
Code changes look good to me and should fix the issue 👍

**Review by asliayk — APPROVED** [review]
tested on ts1, works as expected

**Review by eceeeren — APPROVED** [review]
Tested on TS1, works well!

**Review by MichaelOwenDyer — DISMISSED** [review]
Works as described 👍🏽

**Review by tobias-lippert — APPROVED** [review]
Code LGTM

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 1]
```diff
@@ -185,7 +185,8 @@ public ResponseEntity<Participation> startParticipation(@PathVariable Long exerc
         if (exercise instanceof ProgrammingExercise) {
             // fetch additional objects needed for the startExercise method below
             var programmingExercise = programmingExerciseRepository.findByIdWithTemplateAndSolutionParticipationElseThrow(exercise.getId());
-            if (!featureToggleService.isFeatureEnabled(Feature.ProgrammingExercises) || !isAllowedToParticipateInProgrammingExercise(programmingExercise, null)) {
+            if (!featureToggleService.isFeatureEnabled(Feature.ProgrammingExercises)
+                    || (!authCheckService.isAtLeastInstructorForExercise(exercise, user) && !isAllowedToParticipateInProgrammingExercise(programmingExercise, null))) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:189]: according to the documentation, editors can use this features as well.
https://docs.artemis.cit.tum.de/admin/accessRights/#programming-exercises

Could you please double check this? If this is correct, then we should allow this for editors as well.
Please also adda comment above that shortly explains this in natural language

**Review by krusche — CHANGES_REQUESTED** [review]
see my inline comment

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java (hunk 1 above)
Comment (basak-akan) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:189]: After reviewing the documentation, it's confirmed that editors can use these features too. I will also add a short explanation comment.

**Review by JohannesStoehr — APPROVED** [review]
Latest changes look good to me. Editors can participate as well 👍

**Pull request merged**